llms/mlx_lm/server.py

```diff
@@ -80,344 +75,311 @@ def convert_chat(messages: any, role_mapping: Optional[dict] = None):
         self.end_headers()
-        max_stop_id_sequence_len = (
-            max(len(seq) for seq in stop_id_sequences) if stop_id_sequences else 0
-        )
+
         tokens = []
         current_generated_text_index = 0
-        # Buffer to store the last `max_stop_id_sequence_len` tokens to check for stop conditions before writing to the stream.
+
+        max_stop_id_sequence_len = len(max(stop_id_sequences, default=[]))
+        # Buffer to store the last `max_stop_id_sequence_len` tokens
+        # to check for stop conditions before writing to the stream.
         stop_sequence_buffer = []
-        REPLACEMENT_CHAR = "\ufffd"
+
         for (token, _), _ in zip(
             generate_step(
                 prompt=prompt,
-                model=_model,
-                temp=temperature,
-                top_p=top_p,
-                repetition_penalty=repetition_penalty,
-                repetition_context_size=repetition_context_size,
+                model=MODEL,
+                temp=self.temperature,
+                top_p=self.top_p,
+                repetition_penalty=self.repetition_penalty,
+                repetition_context_size=self.repetition_context_size,
             ),
-            range(max_tokens),
+            range(self.max_tokens),
         ):
             token = token.item()
             tokens.append(token)
             stop_sequence_buffer.append(token)
-            if len(stop_sequence_buffer) > max_stop_id_sequence_len:
-                if REPLACEMENT_CHAR in _tokenizer.decode(token):
-                    continue
-                stop_condition = stopping_criteria(
-                    tokens,
-                    stop_id_sequences,
-                    eos_token_id,
-                )
-                if stop_condition.stop_met:
-                    if stop_condition.trim_length:
-                        tokens = tokens[: -stop_condition.trim_length]
-                    break
-                # This is a workaround because the llama tokenizer emits spaces when decoding token by token.
-                generated_text = _tokenizer.decode(tokens)
-                next_chunk = generated_text[current_generated_text_index:]
-                current_generated_text_index = len(generated_text)
-
-                response = response_creator(response_id, requested_model, next_chunk)
-                try:
-                    self.wfile.write(f"data: {json.dumps(response)}\n\n".encode())
-                    self.wfile.flush()
-                    stop_sequence_buffer = []
-                except Exception as e:
-                    print(e)
-                    break
+
+            # Continue generating tokens until buffer is as large as the longest stop_id_sequence
+            if len(stop_sequence_buffer) < max_stop_id_sequence_len:
+                continue
+
+            # Continue until unicode character is fully generated
+            if "\ufffd" in TOKENIZER.decode(token):
```

Comment:
Would using the variable as `REPLACEMENT_CHAR` make it more readable than `"\ufffd"`?

Reply:
The character only gets used in one place, and in my opinion the comment above it makes it clear enough what it does

Reply:
Not really, readers need to understand that it's a REPLACEMENT_CHAR and then realize that some Unicode characters require more than one character to be decoded correctly. Without this information, it will be very difficult to understand why we need to skip that decoding.

Reply:
How that I think about it, shouldn't we then also remove it from `tokens` and the sequence buffer?

Reply:
We cannot remove it. The REPLACEMENT_CHAR was caused by partial decoding (only the first part of Unicode was generated at the time, and we will need the second token to be able to decode). so, if the next token comes, the tokenizer should be able to properly decode with two tokens.

Reply:
Ah, so it isn't technically needed for decoding of unicode, but the tokenizer is implemented in a way where it looks out for them?

Reply:
Yeah, I'm not an expert on tokenizers, but my understanding is that it has something to do with how tokenizers decode Unicode.

Reply:
Is this enough?
```python
# "\ufffd" is used to indicate to the tokenizer, that subsequent characters
# should be combined into a single unicode character
if "\ufffd" in TOKENIZER.decode(token):
    continue
```